halo2_middleware/src/zal.rs

```diff
@@ -0,0 +1,309 @@
+
+        fn msm_with_cached_scalars(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &[C],
+        ) -> C::Curve;
+
+        fn msm_with_cached_base(
+            &self,
+            coeffs: &[C::Scalar],
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve;
+
+        fn msm_with_cached_inputs(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve;
+        // Execute MSM according to descriptors
+        // Unsure of naming, msm_with_cached_inputs, msm_apply, msm_cached, msm_with_descriptors, ...
+    }
+}
+
+// ZAL using Halo2curves as a backend
+// ---------------------------------------------------
+
+pub mod impls {
+    use std::marker::PhantomData;
+
+    use crate::zal::traits::MsmAccel;
+    use halo2curves::msm::best_multiexp;
+    use halo2curves::CurveAffine;
+
+    // Halo2curve Backend
+    // ---------------------------------------------------
+    #[derive(Default)]
+    pub struct H2cEngine;
+
+    pub struct H2cMsmCoeffsDesc<'c, C: CurveAffine> {
+        raw: &'c [C::Scalar],
+    }
+
+    pub struct H2cMsmBaseDesc<'b, C: CurveAffine> {
+        raw: &'b [C],
+    }
+
+    impl H2cEngine {
+        pub fn new() -> Self {
+            Self {}
+        }
+    }
+
+    impl<C: CurveAffine> MsmAccel<C> for H2cEngine {
+        fn msm(&self, coeffs: &[C::Scalar], bases: &[C]) -> C::Curve {
+            best_multiexp(coeffs, bases)
+        }
+
+        // Caching API
+        // -------------------------------------------------
+
+        type CoeffsDescriptor<'c> = H2cMsmCoeffsDesc<'c, C>;
+        type BaseDescriptor<'b> = H2cMsmBaseDesc<'b, C>;
+
+        fn get_coeffs_descriptor<'c>(&self, coeffs: &'c [C::Scalar]) -> Self::CoeffsDescriptor<'c> {
+            // Do expensive device/library specific preprocessing here
+            Self::CoeffsDescriptor { raw: coeffs }
+        }
+        fn get_base_descriptor<'b>(&self, base: &'b [C]) -> Self::BaseDescriptor<'b> {
+            Self::BaseDescriptor { raw: base }
+        }
+
+        fn msm_with_cached_scalars(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &[C],
+        ) -> C::Curve {
+            best_multiexp(coeffs.raw, base)
+        }
+
+        fn msm_with_cached_base(
+            &self,
+            coeffs: &[C::Scalar],
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve {
+            best_multiexp(coeffs, base.raw)
+        }
+
+        fn msm_with_cached_inputs(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve {
+            best_multiexp(coeffs.raw, base.raw)
+        }
+    }
+
+    // Backend-agnostic engine objects
+    // ---------------------------------------------------
+    #[derive(Debug)]
+    pub struct PlonkEngine<C: CurveAffine, MsmEngine: MsmAccel<C>> {
```

Comment:
I remember we previously discussed two options for this.  Imagining the version with Msm and Fft.
A:
```rust
pub struct PlonkEngine<C: CurveAffine, MsmEngine: MsmAccel<C>, FftEngine: FftAccel<C::Scalar> {
    pub msm_backend: MsmEngine,
    pub fft_backend: FftEngine,
    _marker: PhantomData<C>,
}
```

B:
```rust
pub struct PlonkEngine<C: CurveAffine, MsmEngine: MsmAccel<C>> {
    pub msm_backend: Box<dyn MsmAccel<C>>,
    pub fft_backend: Box<dyn FftAccel<C::Scalar>>,
    _marker: PhantomData<C>,
}
```

In this PR you go with option A.  I would like to review why we go with option A instead of B.

If I remember correctly, trait objects (using `dyn Trait`) don't allow using traits with associated types, which are currently used for the descriptors.  It's very convenient for descriptors to be associated types because it allows us to implement Drop on them, which is a very nice automated way to handle freeing the resources associated with the descriptors.  If we don't have associated types, we either have a generic type for descriptors and free the resources manually, or we make the descriptor type be something like `Box<dyn Descriptor>` and have the Engine do runtime casts to recover the underlying type (which should allow us implementing Drop), but with this option we lose compile time type checking.

Is this review correct?

Reply:
> If I remember correctly, trait objects (using dyn Trait) don't allow using traits with associated types, which are currently used for the descriptors.

Slight correction, they do allow traits with associated types, but each instantiation needs to explicitly mention the associated type, which makes this very cumbersome. We might solve this with an extra macro, but it's extra friction.

> It's very convenient for descriptors to be associated types because it allows us to implement Drop on them, which is a very nice automated way to handle freeing the resources associated with the descriptors. 

Indeed, especially for GPU memory.

>  If we don't have associated types, we either have a generic type for descriptors and free the resources manually, or we make the descriptor type be something like Box<dyn Descriptor> and have the Engine do runtime casts to recover the underlying type (which should allow us implementing Drop), but with this option we lose compile time type checking.

The boxing was also considered to speedup recompilation and code bloat, but since code is open-source, and people are likely to recompile with a single backend, monomorphization bloating was considered not an issue.

We also looked into requiring `Copy` https://github.com/privacy-scaling-explorations/halo2/pull/277#discussion_r1500684988 and https://hackmd.io/F6W1U6X8SLCqTciJt0nfiQ?view but this causes issue with Drop/destructors.